DATAMONGO-2479 - Add AfterConvertCallback and AfterSaveCallback (and …

rpuch opened the pull request:
…their reactive versions).

Previously, only BeforeConvertCallback and BeforeSaveCallback were supported (and their reactive counterparts). This commit adds support for 'after-convert' and 'after-save' events using entity callbacks feature.

Please note that I changed `ReactiveMongoTemplate.DocumentCallback` interface making its only method reactive. It seems ok because it is only used inside reactive contexts (actually, it was only called when doing a `map()` on a `Flux`/`Mono`, so now it is used in `concatMap()`/`flatMap()`).

This supersedes https://github.com/spring-projects/spring-data-mongodb/pull/837 , so I am closing that.

What do you think?

<!--

Thank you for proposing a pull request. This template will guide you through the essential steps necessary for a pull request.
Make sure that:

-->

- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] There is a ticket in the bug tracker for the project in our [JIRA](https://jira.spring.io/browse/DATAMONGO).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

mp911de commented on spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/mapping/event/AfterSaveCallback.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Entity callback triggered after save of a document.
+ *
+ * @author Mark Paluch
```
Feel free to add only your name here, since this is a new file. No need to retain the authors from other types.

rpuch replied:
Fixed

mp911de commented on spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/MongoTemplateUnitTests.java:
```diff
@@ -1921,6 +1931,243 @@ void saveVersionedShouldProjectOnShardKeyWhenLoadingExistingDocument() {
+		verify(afterConvertCallback).onAfterConvert(eq(initialLuke()), eq(document), anyString());
+	}
+
+	@NotNull
```
Nit: No need for `@NotNull`.

rpuch replied:
Removed the annotation

mp911de commented on spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/MongoTemplateUnitTests.java:
```diff
@@ -1921,6 +1931,243 @@ void saveVersionedShouldProjectOnShardKeyWhenLoadingExistingDocument() {
+		));
+	}
+
+	@NotNull
```
Nit: No need for `@NotNull`.

rpuch replied:
Removed the annotation

mp911de commented on spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/ReactiveMongoTemplateUnitTests.java:
```diff
@@ -1096,6 +1109,282 @@ void saveVersionedShouldProjectOnShardKeyWhenLoadingExistingDocument() {
+	}
+
+	@NotNull
+	private Document initialLukeDocument() {
```
Nit: No need for `@NotNull`.

rpuch replied:
Removed the annotation

mp911de commented on spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/ReactiveMongoTemplateUnitTests.java:
```diff
@@ -1096,6 +1109,282 @@ void saveVersionedShouldProjectOnShardKeyWhenLoadingExistingDocument() {
+		));
+	}
+
+	@NotNull
```
Nit: No need for `@NotNull`.

rpuch replied:
Removed the annotation

mp911de commented on spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/UncustomizableFindPublisher.java:
```diff
@@ -0,0 +1,146 @@
+ * @author Roman Puchkovskiy
+ */
+@RequiredArgsConstructor
+class UncustomizableFindPublisher<T> implements FindPublisher<T> {
```
Doesn't it rather make sense to use a proper Mockito mock instead of introducing a `FindPublisher` implementation? This approach requires us to catch up each time MongoDB driver introduces any changes to `FindPublisher`.

rpuch replied:
To be honest, with the number of methods there are on this interface, I thought it would be a nightmare to mock them all. But it looks like all the needed methods are already mocked. I've applied your advice and it seems working ok still being succinct.

rpuch commented:
Thank you for your review, @mp911de ! I've addressed all the issues you mentioned (if I don't miss anything).

As for the performance-related investigation: could you please give me a hint where should I look if you, by chance, already have such a tool (like a benchmark) in the code?

mp911de commented:
We have a few benchmarks here: https://github.com/spring-projects/spring-data-dev-tools/tree/master/benchmark/mongodb

Feel free to have a lot.

rpuch commented:
I've played wth the benchmarks a bit.

First, I had to migrate the benchmarks project to compile with spring-boot 2.3.0 dependencies. I will reference the PR with changes a bit later.

I ran two benchmark sessions. First is against commit 7e3f7bd86137b0d9bd1d4cb3c09ba19bb7674723 on master of spring-data-mongodb (that's the commit to which my commit in branch DATAMONGO-2479 corresponds), so this was the baseline. I'm calling this session 'master', and the corresponding archive is master.tar.gz.

Second session was against commit a54ef345c6547d503fde58070375c004e82d2204 in branch DATAMONGO-2479, I call this session 'branch', the corresponding archive is branch.tar.gz.

From what I see, the results are pretty much the same. For example, here is what I got on master:

```
Benchmark                                        Mode  Cnt        Score       Error  Units
MongoDbBenchmark.convertSingleBook              thrpt   10  2031772,058 ± 27039,347  ops/s
MongoDbBenchmark.findAll                        thrpt   10    13587,815 ±  1000,794  ops/s
MongoDbBenchmark.findByTitle                    thrpt   10    13846,553 ±   757,232  ops/s
MongoDbBenchmark.findByTitleOptional            thrpt   10    12856,971 ±  2289,165  ops/s
MongoDbBenchmark.rawFindAll                     thrpt   10    15628,538 ±  1230,409  ops/s
MongoDbBenchmark.rawFindByTitle                 thrpt   10    15742,024 ±   412,087  ops/s
MongoDbBenchmark.repositoryFindAll              thrpt   10    10783,801 ±  3571,915  ops/s
MongoDbBenchmark.repositoryFindByTitle          thrpt   10    12384,333 ±   799,111  ops/s
MongoDbBenchmark.repositoryFindByTitleDeclared  thrpt   10    13278,882 ±   785,733  ops/s
MongoDbBenchmark.repositoryFindByTitleOptional  thrpt   10     9310,293 ±  3023,870  ops/s
```

and here are the results on the branch:

```
Benchmark                                        Mode  Cnt        Score       Error  Units
MongoDbBenchmark.convertSingleBook              thrpt   10  2045614,403 ± 40567,310  ops/s
MongoDbBenchmark.findAll                        thrpt   10    10952,504 ±   234,477  ops/s
MongoDbBenchmark.findByTitle                    thrpt   10    11276,996 ±  1772,170  ops/s
MongoDbBenchmark.findByTitleOptional            thrpt   10    11642,227 ±  1720,294  ops/s
MongoDbBenchmark.rawFindAll                     thrpt   10    14637,392 ±  1520,145  ops/s
MongoDbBenchmark.rawFindByTitle                 thrpt   10    15816,345 ±  1077,191  ops/s
MongoDbBenchmark.repositoryFindAll              thrpt   10     9705,944 ±   292,123  ops/s
MongoDbBenchmark.repositoryFindByTitle          thrpt   10    10444,036 ±  2058,041  ops/s
MongoDbBenchmark.repositoryFindByTitleDeclared  thrpt   10    13733,597 ±   758,627  ops/s
MongoDbBenchmark.repositoryFindByTitleOptional  thrpt   10    11619,635 ±   773,341  ops/s
```

 I'm attaching the full results (logs and reports) as archives.

[master.tar.gz](https://github.com/spring-projects/spring-data-mongodb/files/4332661/master.tar.gz)
[branch.tar.gz](https://github.com/spring-projects/spring-data-mongodb/files/4332662/branch.tar.gz)

I also added one more benchmark: `AfterConvertCallbacksBenchmark`. It is modeled after `CallbacksBenchmark`.

mp911de commented:
Great work, thank you. There's a slight difference if you check `findAll` (`13587,815` vs. `10952,504`). These differences show only up when returning multiple elements. Scenarios returning just one element look unchanged.

Based on these numbers, it makes sense to merge this PR and we can investigate on our end how we can improve `EntityCallback` performance in Spring Data Commons.

rpuch commented:
Should I rebase on the current master?

mp911de commented:
Feel free, however, there's no need as GitHub does not show any merge conflicts. Please revisit the formatting changes around imports to reduce the noise in the changes. We will take care of the remaining issues during the merge.

rpuch commented:
Then I guess I will leave it as is if the rebase is not mandatory.
I've looked through the import changes once more, seems to be ok.

mp911de commented:
Thank you for your contribution. That's merged and polished now.
